ethereum/spec/src/main/java/tech/pegasys/teku/spec/constants/SpecConstantsReader.java

```diff
@@ -0,0 +1,172 @@
+      CaseFormat.LOWER_CAMEL.converterTo(CaseFormat.UPPER_UNDERSCORE);
+  private final ImmutableMap<Class<?>, Function<Object, ?>> parsers =
+      ImmutableMap.<Class<?>, Function<Object, ?>>builder()
+          .put(Integer.TYPE, this::parseInt)
+          .put(Integer.class, this::parseInt)
+          .put(Long.TYPE, this::parseLong)
+          .put(Long.class, this::parseLong)
+          .put(UInt64.class, fromString(UInt64::valueOf))
+          .put(String.class, Function.identity())
+          .put(Bytes.class, fromString(Bytes::fromHexString))
+          .put(Bytes4.class, fromString(Bytes4::fromHexString))
+          .put(boolean.class, fromString(Boolean::valueOf))
+          .build();
+
+  public SpecConstants read(final InputStream source) throws IOException {
+    final SpecConstantsBuilder constantsBuilder = SpecConstants.builder();
+    final Map<String, Object> rawValues = readValues(source);
+    final Map<String, Object> unprocessedConstants = new HashMap<>(rawValues);
+
+    constantsBuilder.rawConstants(rawValues);
+    streamConstantSetters()
+        .forEach(
+            setter -> {
+              final String constantKey = camelToSnakeCase.convert(setter.getName());
+              final Object rawValue = unprocessedConstants.get(constantKey);
+              invokeSetter(setter, constantsBuilder, constantKey, rawValue);
+              unprocessedConstants.remove(constantKey);
+            });
+
+    if (unprocessedConstants.size() > 0) {
+      final String unknownKeys = String.join(",", unprocessedConstants.keySet());
+      throw new IllegalArgumentException("Detected unknown constants: " + unknownKeys);
+    }
+
+    return constantsBuilder.build();
+  }
+
+  @SuppressWarnings("unchecked")
+  private Map<String, Object> readValues(final InputStream source) throws IOException {
+    final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+    try {
+      final Map<String, Object> values =
+          (Map<String, Object>)
+              mapper
+                  .readerFor(
+                      mapper
+                          .getTypeFactory()
+                          .constructMapType(Map.class, String.class, Object.class))
+                  .readValues(source)
+                  .next();
+      return values;
+    } catch (NoSuchElementException e) {
+      throw new IllegalArgumentException("Supplied constants are empty");
+    }
+  }
+
+  private Stream<Method> streamConstantSetters() {
+    // Ignore any setters that aren't for individual constants
+    final List<String> nonConstantSetters = List.of("rawConstants");
+
+    return Arrays.stream(SpecConstantsBuilder.class.getMethods())
+        .filter(m -> Modifier.isPublic(m.getModifiers()))
+        .filter(m -> m.getReturnType() == SpecConstantsBuilder.class)
+        .filter(m -> m.getParameterTypes().length == 1)
+        .filter(m -> nonConstantSetters.stream().noneMatch(s -> s.equals(m.getName())));
```

Comment:
nit: Make `nonConstantSetters` a `Set` and just use:
```suggestion
        .filter(m -> !nonConstantSetters.contains(m.getName()));
```